Fix panic when fetching user preferences

jakule opened the pull request:
Closes https://github.com/gravitational/teleport/issues/28740

jakule committed:
Fix panic when fetching user preferences

Closes https://github.com/gravitational/teleport/issues/28740

jakule committed:
Prevent overwriting mismatched types in user preferences

zmb3 commented on lib/services/local/userpreferences.go:
```diff
@@ -142,36 +146,71 @@ func createBackendItem(username string, preferences *userpreferencesv1.UserPrefe
+		return
+	}
+
+	srcV := reflect.ValueOf(src).Elem()
```
🙈

jakule replied:
I know :see_no_evil: But IMO doesn't make sense to have the default values already specified above and keep adding nested ifs to set the same values instead of using reflection here. I was hoping to find something in `proto` package, but `proto.Range()` only iterates over set values, and `proto.Merge()` merges collections, so it won't work here.

probakowski replied:
But we do want to iterate only over fields that are set in `src`, right? So it seems `Range()` should do the job?. OTOH `Range` is pretty close to reflection either way so.. 🤷

probakowski replied:
So.. it can be used? AFAICT the flow is as follow:
- find all fields that are set (non-default) in src
- check if field is set in dst
- if not set it to value from src
`Range` can give you the first step, right?

jakule replied:
@zmb3 @probakowski I've removed the reflection explicit from our code.

jakule committed:
Add error handling to user preferences service

jakule committed:
GCI imports

jakule committed:
Refactor user preferences test and logic

This commit updates the approach for comparing equality in user preferences test. It introduces the use of the "go-cmp" library which provides more flexibility in handling comparison of struct elements, thus helping to catch any unexpected changes that could be overlooked with the standard equality checking. We also revised the logic to handle default preferences by overwriting values for better efficiency and readability, also removing unnecessary checking and merging of values.

tigrato commented on lib/services/local/userpreferences.go:
```diff
@@ -142,36 +153,34 @@ func createBackendItem(username string, preferences *userpreferencesv1.UserPrefe
-		dest.Theme = src.Theme
-	}
+// overwriteValues overwrites the values in dst with the values in src.
+func overwriteValues(dst, src protoreflect.ProtoMessage) error {
```
can you please add a comment mentioning that it only works because proto.range only visits non-nil/non-empty fields? 

It would become clear to anyone using it how it works.

jakule replied:
Added

jakule committed:
Add comments to overwriteValues method in userpreferences.go

Added detailed comments to better explain the overwriteValues function in userpreferences.go file. The comments clarify how the function uses proto.Ranges to iterate over fields and only overwrite non-nil/empty fields.

jakule committed:
Apply some magic to preferences test

jakule committed:
Change the import to avoid go.mod changed and match our other imports
